ddtrace/contrib/grpc/client_interceptor.py

```diff
@@ -1,61 +1,82 @@
-        self._host = host
-        self._port = port
-
-    def _start_span(self, method):
-        span = self._pin.tracer.trace('grpc.client', span_type='grpc', service=self._pin.service, resource=method)
-        span.set_tag('grpc.host', self._host)
-        if (self._port is not None):
-            span.set_tag('grpc.port', self._port)
-        if self._pin.tags:
-            span.set_tags(self._pin.tags)
-        # set analytics sample rate
-        span.set_tag(
-            ANALYTICS_SAMPLE_RATE_KEY,
-            config.grpc.get_analytics_sample_rate()
-        )
-        return span
 
-    def intercept_unary_unary(self, continuation, client_call_details, request):
-        return self.intercept_unary_stream(continuation, client_call_details, request)
+def create_client_interceptor(pin):
+    def interceptor_function(continuation, client_call_details,
+                             request_or_iterator):
+        if not pin.enabled:
+            response = continuation(client_call_details, request_or_iterator)
+            return response
+
+        with pin.tracer.trace(
+                '{}.client'.format(pin.app),
+                span_type=config.grpc.span_type,
+                service=pin.service,
+                resource=client_call_details.method
+        ) as span:
+            if pin.tags:
+                span.set_tags(pin.tags)
+
+            span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.grpc.get_analytics_sample_rate())
+
+            headers = {}
+            if config.grpc.distributed_tracing_enabled:
+                propagator = HTTPPropagator()
+                propagator.inject(span.context, headers)
+
+            metadata = []
+            if client_call_details.metadata is not None:
+                metadata = list(client_call_details.metadata)
+            metadata.extend(headers.items())
+
+            client_call_details = _ClientCallDetails(
+                client_call_details.method,
+                client_call_details.timeout,
+                metadata,
+                client_call_details.credentials
+            )
 
-    def intercept_unary_stream(self, continuation, client_call_details, request):
-        if not self._pin or not self._pin.enabled():
-            return continuation(client_call_details, request)
-        with self._start_span(client_call_details.method) as span:
-            new_details = inject_span(span, client_call_details)
             try:
-                return continuation(new_details, request)
+                response = continuation(client_call_details, request_or_iterator)
+                return response
             except Exception:
                 span.set_traceback()
                 raise
 
+    return _ClientInterceptor(interceptor_function)
+
+
+class _ClientCallDetails(
+        collections.namedtuple(
+            '_ClientCallDetails',
+            ('method', 'timeout', 'metadata', 'credentials')),
+        grpc.ClientCallDetails):
```

Comment:
oh whoah, this is interesting and fun... didn't know you could inherit from a `namedtuple`... 0_0